core/src/main/java/org/apache/druid/data/input/impl/RetryingInputStream.java

```diff
@@ -19,99 +19,105 @@
+ * stream for the input object on the given {@link #retryCondition}.
  *
  * @param <T> object type
  */
 public class RetryingInputStream<T> extends InputStream
 {
-
-  public static final Predicate<Throwable> DEFAULT_RETRY_CONDITION = Predicates.alwaysFalse();
-  public static final Predicate<Throwable> DEFAULT_RESET_CONDITION = RetryingInputStream::isConnectionReset;
-
   private static final Logger log = new Logger(RetryingInputStream.class);
 
   private final T object;
   private final ObjectOpenFunction<T> objectOpenFunction;
   private final Predicate<Throwable> retryCondition;
-  private final Predicate<Throwable> resetCondition;
-  private final int maxRetry;
+  private final int maxTries;
 
   private CountingInputStream delegate;
   private long startOffset;
 
+  // Used in tests to disable waiting.
+  private boolean doWait;
+
   /**
-   *
-   * @param object The object entity to open
+   * @param object             The object entity to open
    * @param objectOpenFunction How to open the object
-   * @param retryCondition A predicate on a throwable to indicate if stream should retry. This defaults to
-   *                       {@link IOException}, not retryable, when null is passed
-   * @param resetCondition A predicate on a throwable to indicate if stream should reset. This defaults to
-   *                       a generic reset test, see {@link #isConnectionReset(Throwable)} when null is passed
-   * @param maxRetry      The maximum times to retry. Defaults to {@link RetryUtils#DEFAULT_MAX_TRIES} when null
+   * @param retryCondition     A predicate on a throwable to indicate if stream should retry.
+   * @param maxTries           The maximum times to try. Defaults to {@link RetryUtils#DEFAULT_MAX_TRIES} when null
+   *
    * @throws IOException
    */
   public RetryingInputStream(
       T object,
       ObjectOpenFunction<T> objectOpenFunction,
-      @Nullable Predicate<Throwable> retryCondition,
-      @Nullable Predicate<Throwable> resetCondition,
-      @Nullable Integer maxRetry
+      Predicate<Throwable> retryCondition,
+      @Nullable Integer maxTries
   ) throws IOException
   {
-    this.object = object;
-    this.objectOpenFunction = objectOpenFunction;
-    this.retryCondition = retryCondition == null ? DEFAULT_RETRY_CONDITION : retryCondition;
-    this.resetCondition = resetCondition == null ? DEFAULT_RESET_CONDITION : resetCondition;
-    this.maxRetry = maxRetry == null ? RetryUtils.DEFAULT_MAX_TRIES : maxRetry;
+    this.object = Preconditions.checkNotNull(object, "object");
+    this.objectOpenFunction = Preconditions.checkNotNull(objectOpenFunction, "objectOpenFunction");
+    this.retryCondition = Preconditions.checkNotNull(retryCondition, "retryCondition");
+    this.maxTries = maxTries == null ? RetryUtils.DEFAULT_MAX_TRIES : maxTries;
     this.delegate = new CountingInputStream(objectOpenFunction.open(object));
+    this.doWait = true;
+
+    if (this.maxTries <= 1) {
```

Comment:
Why is maxTries with 1 not allowed? From the implementation below we can see that if maxTries equals to 1, it also reads the underlying stream.

Reply:
Ah, I was thinking it's because if you have only one try, there's no point using this class, because there will never be a retry. You should just use the underlying input stream directly.

Reply:
Hope this comments makes sense; if I missed something I am happy to adjust it in a follow up.